### ISSUE TITLE
[RFC] Per-(model × scenario × prompt) optimization router — unified eligibility selection across PFlash, suffix-decoding, MTP, DFlash, prefix cache, streaming router

### ISSUE THREAD

**Issue by raullenchai** [raullenchai/vllm-mlx#658]
## Motivation

As of v0.7.27 we've shipped a meaningful stack of inference optimizations:

| Technique | Triggering condition | Why it can't be always-on |
|---|---|---|
| **PFlash** (prefill compression, #287/#649) | prompt ≥ ~2.5K tokens, no tool calls, no structured output | short prompt not worth it; tool calls need token integrity |
| **Suffix Decoding** | n-gram match probability high (code/repetitive) | creative writing hits poorly, can be net-slower |
| **MTP** (Multi-Token Prediction) | model has MTP head (Qwen3.5 family) | untrained models can't use it |
| **DFlash** (diffusion spec decode) | long output + high-entropy regions | small models break, short output no benefit |
| **Prefix cache** | multi-turn / agent tool history overlap | single-query is wasted work |
| **Hybrid scheduler** (MoE) | large MoE (122B+) | small dense doesn't need it |
| **Cloud fallback** | local RAM insufficient | privacy-sensitive prompts can't go to cloud |
| **OpenAI Harmony streaming** | Codex CLI `/v1/responses` shape | regular chat doesn't need it |

Eligibility today is encoded as **static per-alias tier annotations** in `vllm_mlx/aliases.json` (e.g. `pflash_tier: "verified"`, `suffix_decoding_tier`, MTP/DFlash eligibility) plus engine-init-time decisions in `vllm_mlx/scheduler.py` and `vllm_mlx/cli.py`. This works but is **static** — same decision for every request that lands on a given alias.

The fact landing on us: **optimization eligibility is genuinely per-(model × scenario × prompt)**, not per-model.

## Concrete problem cases

1. **Same model, different prompt → different optimal stack**
   - `qwen3.5-9b-4bit` on a 50-page report summary: PFlash ON, suffix-decoding OFF (no repetition), MTP ON
   - `qwen3.5-9b-4bit` writing a React component: PFlash OFF (short prompt), suffix-decoding ON (boilerplate hits hard), MTP ON
   - `qwen3.5-9b-4bit` in an agent tool-call loop: PFlash MUST be OFF, prefix cache critical, streaming router matters

2. **Same prompt, state changes mid-request → re-route**
   - Long-document summary phase (PFlash ON) → user follow-up short question with cached prefix (PFlash now wasted) → tool-call enters (PFlash must turn off) → tool result returns (can re-enable)

3. **Cross-technique correctness constraints**
   - PFlash compresses prompt tokens. A `requires_prompt_integrity=True` request (structured output, function calling) must skip PFlash regardless of alias tier.
   - MTP draft acceptance interacts with tool_call parser — known coverage gap.

## Proposal

A small `OptimizationPolicy` layer that runs once per request, **between API ingress and scheduler.add_request**, returning the resolved optimization stack for that specific request.

```python
# Concept sketch — NOT current code
@dataclass(frozen=True)
class RequestSignature:
    token_count: int
    has_tools: bool
    has_structured_output: bool
    repetition_score: float       # cheap n-gram entropy
    estimated_output_length: int  # from max_tokens + stop
    is_followup_short: bool       # short turn after long prior context
    client_shape: str             # "chat", "responses", "codex", ...

@dataclass(frozen=True)
class ResolvedStack:
    pflash: bool
    suffix_decoding: bool
    mtp: bool
    prefix_cache: bool            # almost always True
    stream_router: str

class OptimizationPolicy:
    \"\"\"Per-alias static config + per-request resolver.\"\"\"
    def __init__(self, alias_profile): ...
    def resolve(self, request: Request, sig: RequestSignature) -> ResolvedStack: ...
```

### Phase 1 — Refactor static config into `OptimizationPolicy` (1-2 PRs)

- Lift `pflash_tier` / `suffix_decoding_tier` / `mtp` / `dflash` eligibility from scattered call sites into one `OptimizationPolicy` dataclass built at engine startup from the resolved alias
- Add a no-op `policy.resolve(request, sig)` that returns the same per-alias static decision today — pure refactor, no behavior change
- Adds the seam for Phase 2 without changing latency or test surface

### Phase 2 — Cheap heuristic resolver (1 PR)

- Implement `analyze_prompt(request) → RequestSignature` using stdlib only (token count, regex/grammar for tools, max_tokens for output estimate)
- Wire `policy.resolve()` to use the signature:
  - PFlash: tier-verified AND token_count > breakeven AND not has_tools AND not has_structured_output
  - Suffix decoding: tier-verified AND repetition_score > threshold AND not creative_writing_heuristic
  - Prefix cache: always on (unless PFlash compressed)
  - Stream router: from client_shape
- Adds 1-2 ms per request, removes a few false-on cases users hit today

### Phase 3 — Telemetry collection (1 PR + soak)

- Per-request telemetry: `{alias, signature, resolved_stack, ttft, tps, completion_tokens, accepted_draft_ratio}`
- Local-only by default, opt-in upload (respects existing telemetry plumbing)
- 1-2 months of real workload data → ground truth for which stack actually wins per signature class

### Phase 4 — Data-driven resolver (1 PR)

- Train a tiny classifier (logistic / gradient-boosted tree, NOT a neural network) on collected telemetry
- Replace the heuristic in `policy.resolve()` with the classifier
- A/B-flag-gated rollout so we can compare vs the heuristic

### Phase 5 — Adaptive / mid-stream re-routing (longer term)

- Re-evaluate stack after first generation block when prompt signature changes (tool-call entering, prefix becomes wildly different)
- This is hard — needs scheduler API surface for in-flight reconfiguration
- Only justify if telemetry shows Phase 4 leaves real money on the table

## Why not just add more `pflash_tier`-style fields?

We already do that — and it's the **right Phase 0**. But each new technique (DFlash, future spec-decode variants, future prompt processors) adds a new tier field, scattered enabling code, and another dimension to test. Phase 1's refactor unifies the dispatch site so Phase 2-4 don't multiply the surface area further.

## Non-goals

- Not changing the per-alias static annotations in `aliases.json` — they remain authoritative for "can this technique work at all on this model"
- Not replacing the cloud-fallback policy or hybrid-scheduler routing — those operate on a different axis (resource/cost) and stay where they are
- Not building a generic LLM-based routing layer — overkill, latency-prohibitive

## Open questions

- **Breakeven thresholds** — PFlash breakeven token count varies by model size. Calibrate per alias? Or one global threshold that's slightly conservative?
- **Signature stability** — `repetition_score` on the first 256 tokens of a prompt might mispredict for a long mixed-content prompt. Sample more? Cheaper proxies?
- **Backward compat** — `--pflash always` / `--pflash off` CLI flags should still override the resolver. Same for `--suffix-decoding`. Easy.
- **Test strategy** — golden-input → expected `ResolvedStack` table, plus end-to-end "this stack actually fires" assertions wired into existing bench tiers.

## Related

- #287 / #649 (PFlash; landed in v0.7.27 with `pflash_tier="verified"` for Qwen3.5/3.6)
- #647 / #650 / #654 (R2 mirror; unrelated but same release)
- Existing tier annotations: `vllm_mlx/aliases.json`, `vllm_mlx/model_aliases.py`
- DFlash eligibility check pattern: `vllm_mlx/speculative/dflash/eligibility.py`

## Status

RFC — design conversation, not yet a committed roadmap item. Comments welcome. If consensus forms, Phase 1 is a 1-2 day refactor PR; the bigger investment is Phase 3 telemetry + Phase 4 model training.

**Comment by raullenchai** [comment]
Closing as an unadopted broad RFC, not as completed. The optimization mechanisms have continued to evolve through their own eligibility and policy paths, while this cross-technique router did not reach consensus or a committed roadmap. If a concrete interaction now requires shared routing, please open a focused RFC against the current engine with that specific consumer and measurable acceptance criteria.
